src/sage/matroids/circuits_matroid.pyx

```diff
@@ -0,0 +1,613 @@
+        else:
+            self._groundset = frozenset(groundset)
+            self._C = SetSystem(list(groundset), frozenset([frozenset(C) for C in circuits]))
+        # k-circuits
+        self._k_C = {}
+        for C in self._C:
+            try:
+                self._k_C[len(C)] += [C]
+            except KeyError:
+                self._k_C[len(C)] = []
+                self._k_C[len(C)] += [C]
+        self._matroid_rank = self.rank(self._groundset)
+        self._nsc_defined = nsc_defined
+
+    cpdef groundset(self) noexcept:
+        """
+        Return the groundset of the matroid.
+
+        The groundset is the set of elements that comprise the matroid.
+
+        OUTPUT: a set
+
+        EXAMPLES::
+
+            sage: M = matroids.Theta(2)
+            sage: sorted(M.groundset())
+            ['x0', 'x1', 'y0', 'y1']
+        """
+        return self._groundset
+
+    cpdef _rank(self, X) noexcept:
+        """
+        Return the rank of a set ``X``.
+
+        This method does no checking on ``X``, and ``X`` may be assumed to have
+        the same interface as ``frozenset``.
+
+        INPUT:
+
+        - ``X`` -- an object with Python's ``frozenset`` interface
+
+        OUTPUT: an integer; the rank of ``X`` in the matroid
+
+        EXAMPLES::
+
+            sage: M = matroids.Theta(3)
+            sage: M._rank(['x1', 'y0', 'y2'])
+            2
+        """
+        return len(self._max_independent(X))
+
+    # optional
+
+    cpdef full_rank(self) noexcept:
+        r"""
+        Return the rank of the matroid.
+
+        The *rank* of the matroid is the size of the largest independent
+        subset of the groundset.
+
+        OUTPUT: an integer; the rank of the matroid
+
+        EXAMPLES::
+
+            sage: M = matroids.Theta(20)
+            sage: M.full_rank()
+            20
+        """
+        return self._matroid_rank
+
+    cpdef _is_independent(self, F) noexcept:
+        """
+        Test if input is independent.
+
+        INPUT:
+
+        - ``X`` -- An object with Python's ``frozenset`` interface containing
+          a subset of ``self.groundset()``
+
+        OUTPUT: boolean
+        """
+        I = set(F)
+        s = len(F)
+        for i in self._k_C:
+            if i <= s:
+                for C in self._k_C[i]:
+                    if C <= I:
+                        return False
+        return True
+
+    cpdef _max_independent(self, F) noexcept:
+        """
+        Compute a maximal independent subset.
+
+        INPUT:
+
+        - ``X`` -- An object with Python's ``frozenset`` interface containing
+          a subset of ``self.groundset()``
+
+        OUTPUT: a frozenset; a maximal independent subset of ``X``
+        """
+        I = set(F)
+        for i in self._k_C:
+            for C in self._k_C[i]:
+                if i <= len(I) and i > 0:
+                    if C <= I:
+                        for e in C:
```

Comment:
```suggestion
                        if e in C:
```
otherwise this is not doing anything.

Reply:
This is a fast way of picking an arbitrary element from C. (I know it looks weird.)

Reply:
You should not rely on scoped variables. If you just want to get an element in `C`, then do
```python
e = next(iter(C))
```